storey/flow.py

```diff
@@ -1552,7 +1591,124 @@ def _static_run(*args, **kwargs):
+                mp_context=self._mp_context,
+                initializer=_set_global,
+                initargs=(runnable,),
+            )
+        elif runnable.execution_mechanism == ParallelExecutionMechanisms.thread_pool:
+            self.num_threads += 1
+        elif runnable.execution_mechanism == ParallelExecutionMechanisms.shared_proxy and isinstance(
+            self, ParallelExecution
+        ):
+            self.context.executor.init_runnable(runnable=runnable.name)
+        elif runnable.execution_mechanism not in ParallelExecutionMechanisms.single_thread():
+            raise ValueError(f"Unsupported execution mechanism: {runnable.execution_mechanism}")
+
+    def init_executors(self):
+        # enforce max
+        num_threads = min(self.num_threads * 2, self.max_threads)
+        num_processes = min(self.num_processes * 2, self.max_processes)
+
+        self._executors = {}
+        if num_processes:
+            self._executors[ParallelExecutionMechanisms.process_pool] = ProcessPoolExecutor(
+                max_workers=num_processes, mp_context=self._mp_context
+            )
+        if num_threads:
+            self._executors[ParallelExecutionMechanisms.thread_pool] = ThreadPoolExecutor(max_workers=num_threads)
+
+    def run_executor(
+        self, runnable_name: ParallelExecutionRunnable, runnables_encountered: set[str], event
+    ) -> asyncio.Future:
+        if isinstance(runnable_name, str):
+            runnable = self._runnable_by_name[runnable_name]
+        elif isinstance(runnable_name, ParallelExecutionRunnable):
+            runnable = runnable_name
+        else:
+            raise ValueError("Unexpected runnable type")
+        if id(runnable) in runnables_encountered:
+            raise ValueError(f"select_runnables() returned more than one outlet named '{runnable.name}'")
+        input = (
+            event.body
+            if runnable.execution_mechanism in ParallelExecutionMechanisms.process()
+            else copy.deepcopy(event.body)
+        )
+        runnables_encountered.add(id(runnable))
+        if runnable.execution_mechanism == ParallelExecutionMechanisms.asyncio:
+            future = asyncio.get_running_loop().create_task(runnable._async_run(input, event.path))
+        elif runnable.execution_mechanism == ParallelExecutionMechanisms.naive:
+            future = asyncio.get_running_loop().create_future()
+            future.set_result(runnable._run(input, event.path))
+        elif runnable.execution_mechanism == ParallelExecutionMechanisms.dedicated_process:
+            executor = self._process_executor_by_runnable_name[runnable.name]
+            future = asyncio.get_running_loop().run_in_executor(
+                executor,
+                _static_run,
+                input,
+                event.path,
+            )
+        elif runnable.execution_mechanism == ParallelExecutionMechanisms.shared_proxy and isinstance(
+            self, ParallelExecution
```

Comment:
Maybe we can avoid this awkwardness by using composition over inheritance.

Reply:
I treat it differently please have a look